Add specs for NotificationsController [#311]

tcdowney opened the pull request:
**Issue:** #311 

This includes some minor refactoring of the NotificationsController.

tcdowney commented on app/controllers/notifications_controller.rb:
```diff
@@ -26,33 +31,14 @@ def fetch_notifications
-    # Use callbacks to share common setup or constraints between actions.
-    def set_notification
-      begin
-        @notification = Notification.find(params[:id])
```
This used to let users destroy notifications belonging to other users provided that they knew (or guessed) a valid Notification ID.  I got rid of this `before_action` since it was only used in one place and was duplicating logic and used a more specific query in in `#destroy`.

tcdowney commented on app/controllers/notifications_controller.rb:
```diff
@@ -26,33 +31,14 @@ def fetch_notifications
 
   def signed_in
-    if !user_signed_in?
-      signed_in = -1
```
This code wasn't reachable due to the `if_not_signed_in` before action.  If the intent was for this endpoint to actually return `-1` to un-authenticated users, I can restore this code and prevent the before action from acting on this action.

julianguyen replied:
So this code is being accessed in `notifications.js:197`. This was created to figure out whether the user is signed in from notifications.js.

julianguyen replied:
Your refactor should be fine 👍

tcdowney commented on app/controllers/notifications_controller.rb:
```diff
@@ -1,20 +1,25 @@
     end
   end
 
   def clear
```
```
                          clear_notifications GET      /notifications/clear(.:format)               notifications#clear
    fetch_notifications_notifications GET      /notifications/fetch_notifications(.:format) notifications#fetch_notifications
                   signed_in_notifications GET      /notifications/signed_in(.:format)           notifications#signed_in
                                    notifications GET      /notifications(.:format)                     notifications#index
                                      notification DELETE   /notifications/:id(.:format)                 notifications#destroy
```

I don't think the route for `clear` should be using `HTTP GET` since destroying all notifications for a user is a very stateful request.  It's also susceptible to CSRF attacks since a malicious user could give someone a relatively benign looking link like `http://if-me.org/notifications/clear` and the act of clicking on it (provided they were logged in) would delete all of their notifications whether they liked it or not.

http://programmers.stackexchange.com/questions/188860/why-shouldnt-a-get-request-change-data-on-the-server

Mind if I create an issue for switching this over?

julianguyen replied:
Thanks for pointing this out! Definitely fix this 💯

julianguyen commented:
Looks like @esBeee also made a PR related to the notifications controller (https://github.com/julianguyen/ifme/pull/316). Can you two sync on this? Also @esBeee, please email us at join.ifme@gmail.com to be added to our Slack.

julianguyen commented on app/controllers/notifications_controller.rb:
```diff
@@ -1,20 +1,25 @@
 class NotificationsController < ApplicationController
   before_filter :if_not_signed_in
-  before_action :set_notification, only: [:destroy]
```
I like what @esBeee did to secure notifications https://github.com/julianguyen/ifme/pull/316/files#diff-e7d2878d7ccecd38f6e3a83b5cf85053R73

tcdowney replied:
Cool, glad someone else noticed it earlier!  I'll sync up and work my changes around theirs.  Thanks for the heads up.

julianguyen commented on spec/controllers/notifications_controller_spec.rb:
```diff
@@ -1,3 +1,284 @@
+  describe '#fetch_notifications' do
+    let(:user) { FactoryGirl.create(:user1) }
+    let(:other_user) { FactoryGirl.create(:user2) }
+    let!(:other_user_notification) do
```
Is adding `!` necessary?

tcdowney replied:
It's necessary for this one because I wanted to make sure that we were testing that `fetch_notifications` only returned the current user's notifications.  Since this let is never explicitly invoked anywhere in the test, this second notification won't be created without the `!`.

julianguyen commented:
What's the status on this @tcdowney?

julianguyen commented:
@tcdowney Can you rebase and let me know when it's ready to merge? Thanks :)

tcdowney commented:
@julianguyen 

I rebased this branch with what's on `master` and all tests are passing locally.  Tests are failing on CI, though, but the reason why doesn't seem related to the code that I've changed here -- have you see those failures before? 😢

julianguyen commented:
Not sure, I'm going to rebuild your PR on CI.

julianguyen commented:
@tcdowney 

```
Devise.secret_key was not set. Please add the following to your Devise initializer:

  config.secret_key = '8a920feacbb54c4ee0d68d5f02f556bc57c5d4cc16b3a2048f2651b9c1ad46b20889b4d5c56f2bb330f517166fb610dd2b684ad46652242e7d4576daf6e99a54'
```

Have you generated a secret key and set it to `SECRET_KEY_BASE` in application.yml?

julianguyen commented:
@tcdowney Hmm maybe you need to add yourself to the if me CircleCI org?

brunoocasali commented:
@julianguyen, @tcdowney I don't know what this happens andela-oesho for example, have made a "passed" build... Having an account or not, should not matters to build pass or not, I will rebuild to see.

julianguyen commented:
@brunoocasali Are things working now?

brunoocasali commented:
I can't understand what happened on this PR, but it already merged, we need to merge the #339 to get the new features of continuous integration (by the way there are the changes about C.I.) I suppose when he goes to be merged we have no more problems with it :D
